extras/dynflow-sidekiq.rb

```diff
@@ -25,5 +27,17 @@
   Sidekiq.options[:dynflow_world] = world
 end
 
+# To be able to ensure ordering, dynflow requires that there is only one
+# orchestrator active at the same time.
+# This is enforced by orchestrators contending a lock in redis, calls to
+# dynflow.initialize! block until the lock is acquired by the current process.
+# To play nice with sd_notify, we need to mark the process as ready before it
+# attempts to acquire the lock.
+if Sidekiq.options[:dynflow_executor]
+  Sidekiq::SdNotify.ready
+  Sidekiq::SdNotify.status('orchestrator in passive mode')
+end
```

Comment:
I wonder if we can at least verify a few things. Ideally speaking we would send this **right** before the lock is requested.

Could you help me understand: is the orchestrator a Foreman concept or a Dynflow concept? If it's the former, could/should this notification code live in Dynflow itself? AFAIK Dynflow already has the Sidekiq integration.

Reply:
It is a Dynflow concept, which is only applicable in one specific way of deploying dynflow (when sidekiq is used).

Dynflow already has sidekiq integration, but it is optional. In theory the notification code could live in Dynflow, but we would either have to:
- have sidekiq become a hard dependency (just to use its sdnotify bits in non-sidekiq scenario)
- pull in another dependency for sdnotify
- have disparate feature sets, depending on which deployment model is used.

There are also use cases (like how we run dynflow inside smart proxy), where you might want to defer the notification up to a later stage in the application boot process.

Unless you insist on it, I'd like to treat Dynflow as a library and leave interactions with external systems to its users.

Reply:
> have disparate feature sets, depending on which deployment model is used.

This does feel like the most logical option, though it's not great. On the other hand, you already have that today. Since Sidekiq 6 does have systemd integration already, you gain it regardless. Dynflow already has helpers like `sidekiq_worker?`. Might as well use it.

It does mean you may get more `if` statements though, which increases code complexity.

Another way to see this is that with Sidekiq 6 the executor support will be broken by default when using systemd, requiring every consumer to deal with it.

> Unless you insist on it, I'd like to treat Dynflow as a library and leave interactions with external systems to its users.

I'll trust you on the exact implementation.